uber/site_sections/preregistration.py

```diff
@@ -381,24 +381,27 @@ def register_group_member(self, session, group_id, message='', **params):
                 if not group.unassigned:
                     raise HTTPRedirect('register_group_member?group_id={}&message={}', group_id, 'No more unassigned badges exist in this group')
 
-                badge_being_claimed = group.unassigned[0]
-
-                # Free group badges are only considered 'registered' when they are actually claimed.
+                attrs_to_preserve_from_unassigned_group_member = [
+                    'id',
+                    'group_id',
+                    'badge_type',
+                    'badge_num',
+                    'base_badge_price',
+                    'ribbon',
+                    'paid',
+                    'overridden_price']
+
+                for attr in attrs_to_preserve_from_unassigned_group_member:
+                    if attr in params:
+                        del params[attr]
+
+                attendee = group.unassigned[0]
+                attendee.apply(params, restricted=True)
+
+                # Free group badges are considered registered' when they are actually claimed.
                 if group.cost == 0:
                     attendee.registered = localized_now()
-                else:
-                    attendee.registered = badge_being_claimed.registered
 
-                attendee.badge_type = badge_being_claimed.badge_type
-                attendee.badge_num = badge_being_claimed.badge_num
-                attendee.base_badge_price = badge_being_claimed.base_badge_price
-                attendee.ribbon = badge_being_claimed.ribbon
-                attendee.paid = badge_being_claimed.paid
-                attendee.overridden_price = badge_being_claimed.overridden_price
-
-                session.delete_from_group(badge_being_claimed, group)
```

Comment:
# The Issue

Here is the actual cause of the bug. Imagine you have a group with 4 members, all marked as staff badges. Starts out looking something like this:

Badge Number | Existing Attendees
------------ | --------
\# 10 | Leader
\# 11 | Unassigned A
\# 12 | Unassigned B
\# 13 | Unassigned C

The leader starts assigning badges to her group members. She fills out the form and clicks submit. The newly created attendee will try to claim badge  \# 11. 

Badge Number | Existing Attendees | New Attendees
------------ | -------- | --
\# 10 | Leader |
\# 11 | Unassigned A | New attendee wants to claim \# 11
\# 12 | Unassigned B |
\# 13 | Unassigned C |

So what happens? Well, "Unassigned A" is deleted! And what happens when we delete a badge? We shift all the badges down! So "Unassigned B" gets badge \# 11, AND the new attendee gets badge \# 11!

Badge Number | Existing Attendees | New Attendees
------------ | -------- | --
\# 10 | Leader |
\# 11 | Unassigned B (shifted down) | New attendee wants to claim \# 11
\# 12 | Unassigned C (shifted down) |

The new attendee is saved, and the database complains because both the new attendee and "Unassigned B" have the badge \# 11.

# The Fix

Instead of deleting the existing unassigned attendee, we can simply update its properties with the submitted values! Not only does this fix the bug, it's far more efficient because it avoids the costly badge shifting code!

Win-win!